Index: conf\viewvc.conf.dist
===================================================================
--- conf\viewvc.conf.dist	(revision 2285)
+++ conf\viewvc.conf.dist	(working copy)
@@ -601,6 +601,27 @@
 ##
 #cvsgraph_conf = cvsgraph.conf
 
+# allowed_cvsgraph_useropts: A list of settings used for cvsgraph's
+# user-modifiable behavior which can be changed in the graph display.
+# This value is a comma-delimited list of features, taken from the
+# following set:
+#   invert     - Display graph upside down
+#   branchbox  - Add a branchbox at the tip of each branch
+#   show       - Display user-selected classes of revisions
+#   rotate     - Draw the tree left-to-right or top-to-bottom [*]
+#   limittags  - Allow limit on the maximum number of tags displayed
+#
+# If this option is left unset, users will not be permitted to modify
+# the graph display behavior.
+#
+# [*] WARNING: The 'rotate' option is known to cause some browsers to
+# crash due, presumably, to the display of excessively wide images.
+#
+# Example:
+# allowed_cvsgraph_useropts = limittags, show
+#
+allowed_cvsgraph_useropts = 
+
 ## use_re_search: Enable regular expression search of files in a directory.
 ##
 ## WARNING: Enabling this option can consume HUGE amounts of server
Index: docs\url-reference.html
===================================================================
--- docs\url-reference.html	(revision 2285)
+++ docs\url-reference.html	(working copy)
@@ -654,6 +654,35 @@
     <td>depends</td>
     <td><a href="#root-param"><code>root</code> parameter</a></td>
   </tr>
+  <tr>
+    <td><code>gflip=<var>GFLIP</var></code></td>
+    <td>optional</td>
+    <td>"1" if the revisions in the graph should run
+        youngest-to-oldest; "0" for the reverse</td>
+  </tr>
+  <tr>
+    <td><code>gbbox=<var>GBBOX</var></code></td>
+    <td>optional</td>
+    <td>"1" if the revision graph should contain branch boxes at the
+        tip of each branch; "0" otherwise</td>
+  </tr>
+  <tr>
+    <td><code>gleft=<var>GLEFT</var></code></td>
+    <td>optional</td>
+    <td>"1" if the revision graph should be orientated left-to-right;
+        "0" otherwise</td>
+  </tr>
+  <tr>
+    <td><code>gmaxtag=<var>GMAXTAG</var></code></td>
+    <td>optional</td>
+    <td>maximum number of per-revision tags to show in the revision graph</td>
+  </tr>
+  <tr>
+    <td><code>gshow=<var>GSHOW</var></code></td>
+    <td>optional</td>
+    <td>"all", "inittagged", or "tagged" &mdash; user-selected classes
+        of revision to show in the graph</td>
+  </tr>
 </table>
 
 <h3 id="graphimg-view">Graph Image View</h3>
@@ -700,6 +729,35 @@
     <td>depends</td>
     <td><a href="#root-param"><code>root</code> parameter</a></td>
   </tr>
+  <tr>
+    <td><code>gflip=<var>GFLIP</var></code></td>
+    <td>optional</td>
+    <td>"1" if the revisions in the graph should run
+        youngest-to-oldest; "0" for the reverse</td>
+  </tr>
+  <tr>
+    <td><code>gbbox=<var>GBBOX</var></code></td>
+    <td>optional</td>
+    <td>"1" if the revision graph should contain branch boxes at the
+        tip of each branch; "0" otherwise</td>
+  </tr>
+  <tr>
+    <td><code>gleft=<var>GLEFT</var></code></td>
+    <td>optional</td>
+    <td>"1" if the revision graph should be orientated left-to-right;
+        "0" otherwise</td>
+  </tr>
+  <tr>
+    <td><code>gmaxtag=<var>GMAXTAG</var></code></td>
+    <td>optional</td>
+    <td>maximum number of per-revision tags to show in the revision graph</td>
+  </tr>
+  <tr>
+    <td><code>gshow=<var>GSHOW</var></code></td>
+    <td>optional</td>
+    <td>"all", "inittagged", or "tagged" &mdash; user-selected classes
+        of revision to show in the graph</td>
+  </tr>
 </table>
 
 <h3 id="log-view">Log View</h3>
Index: docs\template-authoring-guide.html
===================================================================
--- docs\template-authoring-guide.html	(revision 2285)
+++ docs\template-authoring-guide.html	(working copy)
@@ -599,6 +599,45 @@
      <a href="#variables-common">COMMON</a> variable set</td>
 </tr>
 <tr class="varlevel1">
+  <td class="varname">gbbox</td>
+  <td>Boolean</td>
+  <td>Toggle generation of a branch box at the tip of all branches in
+      the revision graph.</td>
+</tr>
+<tr class="varlevel1">
+  <td class="varname">gflip</td>
+  <td>Boolean</td>
+  <td>Toggle the direction of the revision graph.</td>
+</tr>
+<tr class="varlevel1">
+  <td class="varname">gleft</td>
+  <td>Boolean</td>
+  <td>Toggle the orientation of the revision graph.</td>
+</tr>
+<tr class="varlevel1">
+  <td class="varname">gmaxtag</td>
+  <td>String</td>
+  <td>Number of tags per revision to display in the revision graph.</td>
+</tr>
+<tr class="varlevel1">
+  <td class="varname">graph_action</td>
+  <td>String</td>
+  <td>Form action URL for the graph customization form.</td>
+</tr>
+<tr class="varlevel1">
+  <td class="varname">graph_hidden_values</td>
+  <td>String</td>
+  <td>Hidden value name/value pairs for the graph customization form.</td>
+</tr>
+<tr class="varlevel1">
+  <td class="varname">gshow</td>
+  <td>String</td>
+  <td>Classes of revisions to show in the revision graph.  Valid values
+      are <tt>all</tt> (all revision), <tt>inittagged</tt> (initial
+      revision(s) and tagged revisions), and <tt>tagged</tt> (tagged
+      revisions only).</td>
+</tr>
+<tr class="varlevel1">
   <td class="varname">imagemap</td>
   <td>String</td>
   <td>HTML markup containing the image map associated with the
@@ -610,6 +649,37 @@
   <td>URL of the ViewVC revision graph image for the current
       resource.</td>
 </tr>
+<tr class="varlevel1">
+  <td class="varname">opt_gbbox</td>
+  <td>Boolean</td>
+  <td>Specifies whether the user is allowed to toggle the generation
+      of branch boxes at the tip of all branches in the revision
+      graph.</td>
+</tr>
+<tr class="varlevel1">
+  <td class="varname">opt_gflip</td>
+  <td>Boolean</td>
+  <td>Specifies whether the user is allowed to toggle the direction
+      of the revision graph.</td>
+</tr>
+<tr class="varlevel1">
+  <td class="varname">opt_gleft</td>
+  <td>Boolean</td>
+  <td>Specifies whether the user is allowed to toggle the orientation
+      of the revision graph.</td>
+</tr>
+<tr class="varlevel1">
+  <td class="varname">opt_gmaxtag</td>
+  <td>Boolean</td>
+  <td>Specifies whether the user is allowed to configure the maximum
+      number of tags per revision show in the revision graph.</td>
+</tr>
+<tr class="varlevel1">
+  <td class="varname">opt_gshow</td>
+  <td>Boolean</td>
+  <td>Specifies whether the user is allowed to configure which
+      classes of revisions are shown in the revision graph.</td>
+</tr>
 </tbody>
 </table>
 </div>
Index: templates\graph.ezt
===================================================================
--- templates\graph.ezt	(revision 2285)
+++ templates\graph.ezt	(working copy)
@@ -15,4 +15,78 @@
   alt="Revisions of [where]" />
 </div>
 
+[define graph_disp_opts][end]
+[if-any opt_gflip][define graph_disp_opts]1[end][end]
+[if-any opt_gbbox][define graph_disp_opts]1[end][end]
+[if-any opt_gleft][define graph_disp_opts]1[end][end]
+
+[define graph_user_opts][end]
+[if-any graph_disp_opts][define graph_user_opts]1[end][end]
+[if-any opt_gshow][define graph_user_opts]1[end][end]
+[if-any opt_gmaxtag][define graph_user_opts]1[end][end]
+
+[if-any graph_user_opts]
+
+<form method="get" action="[graph_action]">
+<div class="vc_graph_form">
+[for graph_hidden_values]<input type="hidden" name="[graph_hidden_values.name]" value="[graph_hidden_values.value]"/>[end]
+
+[is graph_disp_opts "1"]
+<p><strong>Graph display options:</strong></p>
+<table class="auto">
+[if-any opt_gflip]
+  <tr>
+  <td><input id="gflip" type="checkbox" name="gflip" value="1"[if-any gflip] checked="checked"[end] /></td>
+  <td style="text-align: left;"><label for="gflip">Flip the graph</label></td>
+  </tr>
+[end]
+[if-any opt_gbbox]
+  <tr>
+  <td><input id="gbbox" type="checkbox" name="gbbox" value="1"[if-any gbbox] checked="checked"[end] /></td>
+  <td style="text-align: left;"><label for="gbbox">Add a branch box at the tip of each branch</label></td>
+  </tr>
+[end]
+[if-any opt_gleft]
+  <tr>
+  <td><input id="gleft" type="checkbox" name="gleft" value="1"[if-any gleft] checked="checked"[end] /></td>
+  <td style="text-align: left;"><label for="gleft">Draw tree left-to-right (instead of top-to-bottom)</label></td>
+  </tr>
+[end]
+</table>
+[end]
+
+[if-any opt_gshow]
+<p><strong>Revision display options:</strong></p>
+<table class="auto">
+  <tr>
+  <td><input id="gshow_all" type="radio" name="gshow" value="all"[is gshow "all"] checked="checked"[end] /></td>
+  <td style="text-align: left;"><label for="gshow_all">Show all revisions</label></td>
+  </tr>
+  <tr>
+  <td><input id="gshow_inittagged" type="radio" name="gshow" value="inittagged"[is gshow "inittagged"] checked="checked"[end] /></td>
+  <td style="text-align: left;"><label for="gshow_inittagged">Show only tagged and initial untagged revisions</label></td>
+  </tr>
+  <tr>
+  <td><input id="gshow_tagged" type="radio" name="gshow" value="tagged"[is gshow "tagged"] checked="checked"[end] /></td>
+  <td style="text-align: left;"><label for="gshow_tagged">Show only tagged revisions</label></td>
+  </tr>
+</table>
+[end]
+
+[if-any opt_gmaxtag]
+<p><strong>Tag limitation:</strong></p>
+<table class="auto">
+  <tr>
+  <td><input id="gmaxtag" type="text" size="3" name="gmaxtag" value="[gmaxtag]" /></td>
+  <td style="text-align: left;"><label for="gmaxtag">Maximum number of tags to display (0=all)</label></td>
+  </tr>
+</table>
+[end]
+
+<input type="submit" value="Refresh" />
+
+</div>
+</form>
+[end]
+
 [include "include/footer.ezt"]
Index: templates\docroot\styles.css
===================================================================
--- templates\docroot\styles.css	(revision 2285)
+++ templates\docroot\styles.css	(working copy)
@@ -272,3 +272,8 @@
 .vc_query_form {
   background-color: #e6e6e6;
 }
+
+
+/*** Graph Display Form ***/
+.vc_graph_form {
+}
Index: lib\viewvc.py
===================================================================
--- lib\viewvc.py	(revision 2285)
+++ lib\viewvc.py	(working copy)
@@ -690,6 +690,13 @@
   'revision'      : _re_validate_revnum,
   'content-type'  : _validate_mimetype,
 
+  # for cvsgraph
+  'gflip'         : _re_validate_boolint,
+  'gbbox'         : _re_validate_boolint,
+  'gshow'         : _re_validate_alpha,
+  'gleft'         : _re_validate_boolint,
+  'gmaxtag'       : _re_validate_number,
+
   # for query
   'file_match'    : _re_validate_alpha,
   'branch_match'  : _re_validate_alpha,
@@ -2583,6 +2590,59 @@
     copy_stream(fp, server_fp)
   fp.close()
 
+def cvsgraph_make_reqopt(request, cfgname, queryparam, optvalue):
+  # Return a cvsgraph custom option substring bit OPTVALUE based on
+  # CFGNAME's presence in the allowed list of user-configurable
+  # options and QUERYPARAM's presence and boolean interpretation in
+  # the actual request; otherwise, return the empty string for options
+  # that either aren't overridden or aren't allowed to be overridden.
+  
+  if (cfgname in request.cfg.options.allowed_cvsgraph_useropts) \
+     and (int(request.query_dict.get(queryparam, 0))):
+    return optvalue
+  return ''
+
+def cvsgraph_normalize_gshow(request):
+  # Return the effective value of the 'gshow' query parameter, noting
+  # that a missing parameter is the same as gshow=all, and treating a
+  # bogus parameter value as the same as gshow=all, too.
+  gshow = request.query_dict.get('gshow', 'all')
+  if gshow not in ('all', 'inittagged', 'tagged'):
+    gshow = 'all'
+  return gshow
+  
+def cvsgraph_extraopts(request):
+  # Build a set of -O options for controlling cvsgraph's behavior,
+  # based on what the user has requested and filtered against what the
+  # user is allowed to request.
+  
+  cfg = request.cfg
+
+  ep = '-O'
+
+  # Simple mappings of boolean flags
+  ep = ep + cvsgraph_make_reqopt(request, 'invert', 'gflip',
+                                 ';upside_down=true')
+  ep = ep + cvsgraph_make_reqopt(request, 'branchbox', 'gbbox',
+                                 ';branch_dupbox=true')
+  ep = ep + cvsgraph_make_reqopt(request, 'rotate', 'gleft',
+                                 ';left_right=true')
+
+  # Stripping is a little more complex.
+  if ('show' in request.cfg.options.allowed_cvsgraph_useropts):
+    gshow = cvsgraph_normalize_gshow(request)
+    if gshow == 'inittagged':
+      ep = ep + ';strip_untagged=true'
+    elif gshow == 'tagged':
+      ep = ep + ';strip_untagged=true;strip_first_rev=true'
+
+  # And tag limitation has a user-supplied value to mess with.
+  if ('limittags' in request.cfg.options.allowed_cvsgraph_useropts) \
+     and request.query_dict.has_key('gmaxtag'):
+    ep = ep + ';rev_maxtags=' + request.query_dict['gmaxtag']
+
+  return ep + ';'
+  
 def view_cvsgraph_image(request):
   "output the image rendered by cvsgraph"
   # this function is derived from cgi/cvsgraphmkimg.cgi
@@ -2600,6 +2660,7 @@
   fp = popen.popen(cfg.utilities.cvsgraph or 'cvsgraph',
                    ("-c", cfg.path(cfg.options.cvsgraph_conf),
                     "-r", request.repos.rootpath,
+                    cvsgraph_extraopts(request),
                     rcsfile), 'rb', 0)
   
   copy_stream(fp, get_writeready_server_file(request, 'image/png'))
@@ -2642,12 +2703,28 @@
                                           pathtype=vclib.DIR,
                                           params={'pathrev': None},
                                           escape=1, partial=1),
+                    cvsgraph_extraopts(request),
                     rcsfile), 'rb', 0)
 
+  graph_action, graph_hidden_values = \
+    request.get_form(view_func=view_cvsgraph, params={})
+
   data = common_template_data(request)
   data.merge(ezt.TemplateData({
     'imagemap' : fp,
     'imagesrc' : imagesrc,
+    'graph_action' : graph_action,
+    'graph_hidden_values' : graph_hidden_values,
+    'opt_gflip' : ezt.boolean('invert' in cfg.options.allowed_cvsgraph_useropts),
+    'opt_gbbox' : ezt.boolean('branchbox' in cfg.options.allowed_cvsgraph_useropts),
+    'opt_gshow' : ezt.boolean('show' in cfg.options.allowed_cvsgraph_useropts),
+    'opt_gleft' : ezt.boolean('rotate' in cfg.options.allowed_cvsgraph_useropts),
+    'opt_gmaxtag' : ezt.boolean('limittags' in cfg.options.allowed_cvsgraph_useropts),
+    'gflip' : ezt.boolean(int(request.query_dict.get('gflip', 0))),
+    'gbbox' : ezt.boolean(int(request.query_dict.get('gbbox', 0))),
+    'gleft' : ezt.boolean(int(request.query_dict.get('gleft', 0))),
+    'gmaxtag' : request.query_dict.get('gmaxtag', 0),
+    'gshow' : cvsgraph_normalize_gshow(request),
     }))
   generate_page(request, "graph", data)
 
Index: lib\config.py
===================================================================
--- lib\config.py	(revision 2285)
+++ lib\config.py	(working copy)
@@ -425,6 +425,7 @@
     self.options.detect_encoding = 0
     self.options.use_cvsgraph = 0
     self.options.cvsgraph_conf = "cvsgraph.conf"
+    self.options.allowed_cvsgraph_useropts = []
     self.options.use_re_search = 0
     self.options.dir_pagesize = 0
     self.options.log_pagesize = 0
Index: CHANGES
===================================================================
--- CHANGES	(revision 2285)
+++ CHANGES	(working copy)
@@ -1,3 +1,5 @@
+  * allow user-configurable cvsgraph display (issue #336)
+
 Version 1.1.5 (released 29-Mar-2010)
 
   * security fix: escape user-provided search_re input to avoid XSS attack
